enhancements/network/service-egress-traffic-steering.md

```diff
@@ -0,0 +1,284 @@
+---
+title: service-egress-traffic-steering
+authors:
+  - "@oribon"
+reviewers:
+  - "@trozet"
+approvers:
+  - "@danwinship"
+  - "@trozet"
+api-approvers:
+  - "@trozet"
+creation-date: 2022-07-28
+last-updated: 2022-07-28
+status: implementable
+tracking-link:
+  - https://issues.redhat.com/browse/SDN-2682
+---
+
+# OVN service egress traffic steering
+
+## Summary
+
+Some external systems that communicate with applications running on the Kubernetes cluster through a LoadBalancer service expect that the source IP of traffic originating from the pods backing the service is identical to the destination IP they use to reach them - i.e the LoadBalancer's external IP (the ingress IP). 
+In addition, traffic to and from these pods should be able to use any network interface on the host.
+
+By annotating a LoadBalancer service, users could request that the source IP of packets originating from all of the pods that are endpoints of it would be its external IP. They could also specify from which network the pods' traffic should exit from.
+This mechanism will be supported only on clusters using OVN-Kubernetes on "Local Gateway Mode".
```

Comment:
As the feature relies on the host's routes/iptables I don't see why it's weird, I thought that's the reason we have the two modes? iiuc at least for now we need to support this on local mode only

Reply:
Part of the overall solution is to allow a user to decide which interface the egress packet should go out of. We originally thought that was intertwined with this enhancement, because we thought the user would provide a nexthop and interface for each service in the annotation (to indicate where to send the packet). Now we realize this is not the case and all we need to do is SNAT the packet and let the user's routes decide its fate in the kernel.

With shared gateway mode, we could still do the same thing by forwarding packets "locally" from these endpoints to mp0 and letting the kernel route them. On the other hand we could make shared gw version use OVN SNAT on GR to SNAT the packet to hte service IP and send it directly out of br-ex. I'm not sure how useful that is. I'm also not sure if we want to have 2 different implementations of the same feature (one for shared, one for local gw).

All of that being said I think maybe we should just allow the feature to work the same for both gateway modes (via mp0 and let the kernel route it, SNAT with iptables). Thoughts @danwinship @oribon ?

Reply:
I'm not really up to date on the whole local-vs-shared gateway situation so I don't have a strong opinion. It just seemed weird to me that we were adding a new feature that only works if you enable a non-preferred(?) mode

Reply:
iiuc the current way will work for both shared/local (also did a small check that verifies it). if it makes sense to keep it that way for shared and we aren't missing anything that might block it we can just drop this restriction considering the change of scope.

Reply:
removed mentions of local/shared restrictions